### PR TITLE
[RLlib] Fix synching of global_timestep across all policies when restoring from checkpoint.

### PR DESCRIPTION
**Pull request by sven1977**
Signed-off-by: sven1977 <svenmika1977@gmail.com>

Fix synching of global_timestep across all policies when restoring from checkpoint.

Currently, the counters (from which all Policies' `self.global_timestep` are derived) of an algorithm are NOT restored when restoring from a checkpoint.

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by gjoliver** [comment]
I thought global variables are restored here?
https://github.com/ray-project/ray/blob/master/rllib/evaluation/rollout_worker.py#L1626-L1627

Ideally, we don't want to write restoration logics for specific fields, in case folks add more data, etc.

File: rllib/algorithms/algorithm.py [hunk 1]
```diff
@@ -2496,12 +2496,29 @@ def __getstate__(self) -> dict:
...
+        global_vars = {}
+        if "counters" in state:
+            # Don't recreate `self._counters` (it's already a defaultdict),
+            # just set its values.
+            for key, value in state["counters"].items():
+                self._counters[key] = value
+
+            # Set global vars.
+            if hasattr(self, "workers"):
+                global_vars = {
```
Comment (avnishn) [inline rllib/algorithms/algorithm.py:2518]: maybe i'm not seeing it here, but where does global_vars get used?

File: rllib/evaluation/worker_set.py [hunk 2]
```diff
@@ -225,7 +225,8 @@ def sync_weights(
         if from_worker is not None and self.local_worker() is not None:
             self.local_worker().set_weights(weights, global_vars=global_vars)
         # If `global_vars` is provided and local worker exists  -> Update its
-        # global_vars.
+        # global_vars (global vars of the remote workers have already been
+        # updated above in this case).
         elif self.local_worker() is not None and global_vars is not None:
             self.local_worker().set_global_vars(global_vars)
```
Comment (avnishn) [inline rllib/evaluation/worker_set.py:231]: oh its over here. That's nice.

**Review by avnishn — APPROVED** [review]
yep lgtm

File: rllib/algorithms/algorithm.py (hunk 1 above)
Comment (gjoliver) [reply rllib/algorithms/algorithm.py:2518]: actually more specifically, I don't know why we should construct a global_vars section ourselves ...

**Comment by sven1977** [comment]
Hey @gjoliver , I kind of agree, but this should be a different PR. This PR is just a quick fix for an obvious bug. The more thorough, deeper change would be, imo:
* global_vars should be user-only.
* hence, `global_timestep` should be removed from global_vars
* instead, always synch all of the Algorithm's `self._counters` to all policies (besides global_vars).

Before we do this though, we should find out, whether anyone is even using global_vars. It seems to me a simple override of get_state/set_state in a child Policy would suffice and global_vars is actually completely superfluous.

Thanks for reviewing @avnishn !

**Comment by gjoliver** [comment]
> Hey @gjoliver , I kind of agree, but this should be a different PR. This PR is just a quick fix for an obvious bug. The more thorough, deeper change would be, imo:
> 
> * global_vars should be user-only.
> * hence, `global_timestep` should be removed from global_vars
> * instead, always synch all of the Algorithm's `self._counters` to all policies (besides global_vars).
> 
> Before we do this though, we should find out, whether anyone is even using global_vars. It seems to me a simple override of get_state/set_state in a child Policy would suffice and global_vars is actually completely superfluous.
> 
> Thanks for reviewing @avnishn !

do I understand correctly that you are saying there is a good chance we can get rid of global_vars and its syncing logics al-together?

**Pull request closed without merging**